src/helpers/buffers/unordered_receiver.rs

```diff
@@ -204,29 +209,54 @@ where
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
-            self.wake_next();
-            return Poll::Ready(Ok(m));
-        }
-
-        loop {
-            match self.stream.as_mut().poll_next(cx) {
-                Poll::Pending => {
-                    return Poll::Pending;
-                }
-                Poll::Ready(Some(b)) => {
-                    if let Some(m) = self.spare.extend(b.as_ref()) {
-                        self.wake_next();
-                        return Poll::Ready(Ok(m));
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+
+        let message = self.spare.read::<M>();
+        let next = match message {
+            Some(m) => {
+                // this check exists to make sure the inner stream is eventually moved to
+                // the closed state. We don't want to poll it too often, but we also need to know
+                // when it is done and `UnorderedReceiver` can be dropped.
+                if self.spare.is_empty() {
```

Comment:
This will fail to capture the case where you have an incomplete message buffered.  In that case, you will get `None` from an attempt to read, but you won't have added more data to the buffer.

Because this function cannot know how big the next message is, you can't really do anything different here, but I would argue that topping up the spare is not helpful.

Reply:
This is the intended behavior. If there is something left in the spare, there must be a receive request for it later. This will correctly poll the stream if spare does not have enough bytes. The only edge case, where we don't move the fused stream to the closed state, is when spare is empty after the last read.